src/components/action-bar/action-bar.tsx

```diff
@@ -0,0 +1,204 @@
+     */
+    @Event()
+    public itemSelected: EventEmitter<ActionBarItem>;
+
+    @Element()
+    private host: HTMLElement;
+
+    @State()
+    private overflowCutoff: number = this.actions.length;
+
+    private intersectionObserver: IntersectionObserver;
+    private firstRender = true;
+
+    public render() {
+        let overflowActions: Array<MenuItem | ListSeparator> = [];
+        if (this.actions.length) {
+            overflowActions = this.actions.slice(this.overflowCutoff);
+        }
+
+        return (
+            <Host
+                aria-label={this.accessibleLabel}
+                class={{
+                    'is-full-width': this.layout === 'fullWidth',
+                    'is-floating': this.layout === 'floating',
+                }}
+            >
+                <div class="items">
+                    {this.actions.map(this.renderActionBarItem)}
+                </div>
+                {this.renderOverflowMenu(overflowActions)}
+            </Host>
+        );
+    }
+
+    public componentDidLoad() {
+        this.createIntersectionObserver();
+    }
+
+    public disconnectedCallback() {
+        this.intersectionObserver?.disconnect();
+    }
+
+    private renderActionBarItem = (item: ActionBarItem, index: number) => {
+        return (
+            <limel-action-bar-item
+                item={item}
+                onSelect={this.handleSelect}
+                isVisible={this.isVisible(index)}
+            />
+        );
+    };
+
+    private renderOverflowMenu = (items: Array<MenuItem | ListSeparator>) => {
+        if (!(this.actions.length - this.overflowCutoff)) {
+            return;
+        }
+
+        return (
+            <limel-action-bar-overflow-menu
+                openDirection={this.openDirection}
+                items={items}
+                onSelect={this.handleSelect}
+            />
+        );
+    };
+
+    private isVisible(index: number) {
+        return index < this.overflowCutoff;
+    }
+
+    private handleSelect = (
+        event: CustomEvent<ActionBarItem | ListSeparator>
+    ) => {
+        event.stopPropagation();
+        if (isItem(event.detail)) {
+            this.itemSelected.emit(event.detail);
+        }
+    };
+
+    private handleIntersectionHost = (entries: IntersectionObserverEntry[]) => {
+        const intersectingItems = entries.filter(
+            (entry) => entry.isIntersecting
+        );
+
+        const notIntersectingItems = entries.filter(
+            (entry) => !entry.isIntersecting
+        );
+
+        if (this.firstRender) {
+            this.overflowCutoff = intersectingItems.length;
+        } else {
+            this.overflowCutoff =
+                this.overflowCutoff +
+                intersectingItems.length -
+                notIntersectingItems.length;
+        }
+
+        this.firstRender = false;
+    };
+
+    private createIntersectionObserver() {
+        const options = {
+            root: this.host.parentElement,
```

Comment:
Didn't you want to handle being rendered inside a shadow dom?

```suggestion
            root: this.host.parentNode instanceof ShadowRoot
                ? this.host.parentNode.host
                : this.host.parentElement,
```